cpp/src/plasma/io/connection.cc

```diff
@@ -0,0 +1,327 @@
+    ARROW_LOG(WARNING) << "[" << debug_label_ << "]ProcessMessage with type "
+                       << read_type_ << " took " << interval.count() << " ms.";
+  }
+}
+
+void ClientConnection::ProcessError(const std::error_code& ec) {
+  ARROW_LOG(ERROR)
+      << "Failed when processing message. Disconnecting the client. Error code = " << ec;
+  // If there was an error, disconnect the client.
+  PlasmaConnection::Close();
+}
+
+void ClientConnection::ProcessMessage(int64_t type, int64_t length, const uint8_t* data) {
+  message_handler_(shared_from_this(), type, length, data);
+}
+
+struct AsyncObjectNotificationWriteBuffer : public AsyncWriteBuffer {
+  static std::unique_ptr<AsyncObjectNotificationWriteBuffer> MakeDeletion(
+      const ObjectID& object_id) {
+    auto message = new std::vector<uint8_t>();
+    SerializeObjectDeletionNotification(object_id, message);
+    return std::unique_ptr<AsyncObjectNotificationWriteBuffer>(
+        new AsyncObjectNotificationWriteBuffer(message));
+  }
+
+  static std::unique_ptr<AsyncObjectNotificationWriteBuffer> MakeReady(
+      const ObjectID& object_id, const ObjectTableEntry& entry) {
+    auto message = new std::vector<uint8_t>();
+    SerializeObjectSealedNotification(object_id, entry, message);
+    return std::unique_ptr<AsyncObjectNotificationWriteBuffer>(
+        new AsyncObjectNotificationWriteBuffer(message));
+  }
+
+  void ToBuffers(std::vector<asio::const_buffer>& message_buffers) override {
+    message_buffers.push_back(asio::buffer(&size, sizeof(size)));
+    message_buffers.push_back(asio::buffer(*notification_msg));
+  }
+
+  std::unique_ptr<std::vector<uint8_t>> notification_msg;
+  int64_t size;
+
+ protected:
+  explicit AsyncObjectNotificationWriteBuffer(std::vector<uint8_t>* message) {
+    // Serialize the object.
+    notification_msg.reset(message);
+    size = message->size();
+    handler = [](const asio::error_code& status) {
+      auto errno_ = status.value();
+      if (!errno_) {
+        return;
+      }
+      if (errno_ == EAGAIN || errno_ == EWOULDBLOCK || errno_ == EINTR) {
+        ARROW_LOG(DEBUG) << "The socket's send buffer is full, so we are caching this "
+                            "notification and will send it later.";
+        ARROW_LOG(WARNING) << "Blocked unexpectly when sending message async.";
+      } else {
+        ARROW_LOG(WARNING) << "Failed to send notification to client.";
+        if (errno_ == EPIPE) {
+          // TODO(suquark): We could probably close the socket here.
+        }
+      }
+    };
+  }
+};
+
+Status ClientConnection::SendFd(int fd) {
+  ARROW_CHECK(send_fd(GetNativeHandle(), fd));
+  return Status::OK();
+}
+
+void ClientConnection::SendObjectDeletionAsync(const ObjectID& object_id) {
+  auto raw_ptr = AsyncObjectNotificationWriteBuffer::MakeDeletion(object_id).release();
+  auto write_buffer =
+      std::unique_ptr<io::AsyncWriteBuffer>(static_cast<io::AsyncWriteBuffer*>(raw_ptr));
```

Comment:
is it possible to use std::move here?

Reply:
I am not so sure if `std::move` is necessary here. I seems that the right side has been a rvalue.